### PR TITLE
:sparkles: custom user bind and ephemeral mounts

### PR DESCRIPTION
**Pull request by oz123**
Users can now specify custom and ephemeral mounts in cloud-init under the `install` section, e.g.:

```
users:
 - name: kairos
...
install:
  auto: true
  device: "auto"
  bind_mounts:
  - /mnt/bind1
  - /mnt/bind2
  ephemeral_mounts:
  - /mnt/ephemeral
  - /mnt/ephemeral2
...
```
Ephemeral mounts are mounted as RW - but changes are discarded when the machine is restart.
Bind mounts will persist changes after restarted.

This is a fix for #210

Signed-off-by: Oz Tiram <oz@spectrocloud.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by oz123** [comment]
This is now working. However, the tests are not complete yet and docs are needed.

File: overlay/files/system/oem/11_persistency.yaml [hunk 1]
```diff
@@ -36,6 +36,23 @@ stages:
           /usr/share/pki/trust/anchors
           /var/lib/ca-certificates
         PERSISTENT_STATE_BIND: "true"
+    - if: |
+        [ -r /run/cos/custom-layout.env ]
+      name: "append custom bind and ephemeral mounts to /run/cos/cos-layout.env"
+      commands:
+        - |
+          if [ -r /run/cos/custom-layout.env ]; then
```
Comment (mudler) [inline overlay/files/system/oem/11_persistency.yaml:44]: nit: no need for the double if here

File: tests/custom_mount_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/spectrocloud/peg/matcher"
+)
+
+var _ = Describe("kairos test custom user mounts", Label("custom-mounts-test"), func() {
```
Comment (mudler) [inline tests/custom_mount_test.go:12]: this adds a new test suite, however, as-is won't be called by our runners, care to add it?

Comment (oz123) [reply tests/custom_mount_test.go]: done.

File: overlay/files/system/oem/11_persistency.yaml (hunk 1 above)
Comment (oz123) [reply overlay/files/system/oem/11_persistency.yaml]: done.

File: tests/custom_mount_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/spectrocloud/peg/matcher"
+)
+
+var _ = Describe("kairos test custom user mounts", Label("install-test"), func() {
```
Comment (mudler) [inline tests/custom_mount_test.go:12]:
```suggestion
var _ = Describe("kairos test custom user mounts", Label("custom-mount-test"), func() {
```

File: Earthfile [hunk 4]
```diff
@@ -475,6 +475,39 @@ run-qemu-datasource-tests:
 
     RUN PATH=$PATH:$GOPATH/bin ginkgo --label-filter "$TEST_SUITE" --fail-fast -r ./tests/
 
+run-qemu-install-tests:
+    FROM opensuse/leap
+    WORKDIR /test
+    RUN zypper in -y qemu-x86 qemu-arm qemu-tools go git
+    ARG FLAVOR
+    ARG TEST_SUITE=install-test
```
Comment (mudler) [inline Earthfile:483]:
```suggestion
    ARG TEST_SUITE=custom-mount-test
```

File: Earthfile [hunk 5]
```diff
@@ -475,6 +475,39 @@ run-qemu-datasource-tests:
 
     RUN PATH=$PATH:$GOPATH/bin ginkgo --label-filter "$TEST_SUITE" --fail-fast -r ./tests/
 
+run-qemu-install-tests:
```
Comment (mudler) [inline Earthfile:478]:
```suggestion
run-qemu-custom-mount-tests:
```

File: .github/workflows/image.yaml [hunk 6]
```diff
@@ -180,6 +180,26 @@ jobs:
             ./earthly.sh +datasource-iso --CLOUD_CONFIG=tests/assets/autoinstall.yaml
             ./earthly.sh +run-qemu-datasource-tests --FLAVOR=${{ matrix.flavor }}
 
+  qemu-install-tests:
```
Comment (mudler) [inline .github/workflows/image.yaml:183]:
```suggestion
  qemu-custom-mount-tests:
```

File: .github/workflows/image.yaml [hunk 7]
```diff
@@ -180,6 +180,26 @@ jobs:
...
+         - flavor: "opensuse-leap"
+         - flavor: "ubuntu"
+    steps:
+    - uses: actions/checkout@v3
+    - name: Download artifacts
+      uses: actions/download-artifact@v3
+      with:
+          name: kairos-${{ matrix.flavor }}.iso.zip
+    - run: |
+            ./earthly.sh +run-qemu-install-tests --FLAVOR=${{ matrix.flavor }}
```
Comment (mudler) [inline .github/workflows/image.yaml:201]:
```suggestion
            ./earthly.sh +run-qemu-custom-mount-tests --FLAVOR=${{ matrix.flavor }}
```

File: .github/workflows/image.yaml [hunk 8]
```diff
@@ -180,6 +180,26 @@ jobs:
...
 
+  qemu-install-tests:
+    needs:
+    - build
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+       include:
+         - flavor: "alpine-opensuse-leap"
```
Comment (mudler) [inline .github/workflows/image.yaml:191]: maybe just focus on one flavor only? I don't see anything that should impact this at distro-level (and would also keep a lot of cycles of our CI busy)

File: .github/workflows/image.yaml (hunk 6 above)
Comment (mudler) [reply .github/workflows/image.yaml:183]: do we need to duplicate here, or we can re-use a target that already exists?

File: .github/workflows/image.yaml (hunk 8 above)
Comment (oz123) [reply .github/workflows/image.yaml:191]: Sure, we already removed a lot. I agree it's enough to use only one favor.

**Comment by oz123** [comment]
@mudler I believe this PR can be merged now. The custom-mount-tests are passing, also the other acceptance tests
are passing, except bundle and upgrade tests failing. Their failures seem unrelated to the changes we added (this failures
happen very often in other PRs too).

**Comment by mudler** [comment]
https://github.com/kairos-io/kairos/actions/runs/4029252749/jobs/6927273668 https://github.com/kairos-io/kairos/actions/runs/4029252749/jobs/6927273719 the upgrade test seems to fail consistently

**Comment by mudler** [comment]
needs also docs updates

**Review by mudler — DISMISSED** [review]
looks good here! however tests seems to fail: https://github.com/kairos-io/kairos/actions/runs/4033543193/jobs/6934607749

**Comment by oz123** [comment]
> https://github.com/kairos-io/kairos/actions/runs/4029252749/jobs/6927273668 https://github.com/kairos-io/kairos/actions/runs/4029252749/jobs/6927273719 the upgrade test seems to fail consistently

@mudler seems that was because I added earthlyignore file. I took this commit out, this is now handled in #794 .

**Pull request merged**